benchmarks/transactions.cpp

```diff
@@ -0,0 +1,138 @@
+                                         init_count * 2
+                                             - (w1.balance() + w2.balance()));
+        w1.confirm_transaction(mint_tx);
+    }
+    if(w2.balance() > 0) {
+        m_valid_tx = w2.send_to(w2.balance(), w1.generate_key(), true).value();
+        w1.confirm_transaction(m_valid_tx);
+        w2.confirm_transaction(m_valid_tx);
+    }
+    if(w1.count() != init_count) {
+        // generate an N-1 transaction
+        m_valid_tx = w1.send_to(w1.balance(), w1.generate_key(), true).value();
+        w1.confirm_transaction(m_valid_tx);
+        // fan to 10 (value-2 UTXOs)
+        m_valid_tx = w1.fan(init_count, 2, w1.generate_key(), true).value();
+        w1.confirm_transaction(m_valid_tx);
+    }
+}
+
+/// @brief Time an N-in, 1-out transaction.
+/// @brief Note: handles benchmark timing, do not time outside function.
+/// @param sender
+/// @param reciever
+/// @param n_in
+/// @param state
+/// @return
+inline bool generate_Nto1_tx(cbdc::transaction::wallet& sender,
+                             cbdc::transaction::wallet& reciever,
+                             uint32_t n_in,
+                             benchmark::State& state) {
+    std::optional<cbdc::transaction::full_tx> maybe_tx{};
+    state.ResumeTiming();
+    maybe_tx = sender.send_to(n_in * 2, reciever.generate_key(), true).value();
+    state.PauseTiming();
+    if(maybe_tx.has_value()) {
+        sender.confirm_transaction(*maybe_tx);
+        reciever.confirm_transaction(*maybe_tx);
+        return true;
+    }
+    return false;
+}
+
+/// @brief Time an N-in, 2-out transaction.
+/// @brief Note: handles benchmark timing, do not time outside function.
+/// @param sender
+/// @param reciever
+/// @param n_in
+/// @param state
+/// @return
+inline bool generate_Nto2_tx(cbdc::transaction::wallet& sender,
+                             cbdc::transaction::wallet& reciever,
+                             uint32_t n_in,
+                             benchmark::State& state) {
+    std::optional<cbdc::transaction::full_tx> maybe_tx{};
+    state.ResumeTiming();
+    maybe_tx
+        = sender.send_to(n_in * 2 - 1, reciever.generate_key(), true).value();
+    state.PauseTiming();
+    if(maybe_tx.has_value()) {
+        sender.confirm_transaction(*maybe_tx);
+        reciever.confirm_transaction(*maybe_tx);
+        return true;
+    }
+    return false;
+}
+
+// Benchmarkable N in 1 out transaction (for sweep)
+static void Nto1_tx(benchmark::State& state) {
+    cbdc::transaction::wallet wallet_a;
+    cbdc::transaction::wallet wallet_b;
+    bool valid;
+    for(auto _ : state) {
+        reset_wallets(wallet_a, wallet_b, SWEEP_MAX);
+        valid = generate_Nto1_tx(wallet_a, wallet_b, state.range(0), state);
+        if(!valid) {
+            GTEST_LOG_(ERROR) << state.range(0) << "-2 transaction invalid";
+            return;
+        }
+        ASSERT_EQ(wallet_a.balance(), SWEEP_MAX * 2 - 2 * state.range(0));
```

Comment:
It looks like we're timing everything in these few functions, including the `ASSERT`s; is that something the `BENCHMARK()` pattern is smart enough to ignore?

Reply:
In these functions, starting/stopping the timer is handled by `generate_Nto1_tx()` and `generate_Nto2_tx()`, so the `ASSERT`s are ignored (all timing occurs within these functions, and does not occur within `Nto1_tx(benchmark::State&)` or `Nto2_tx(benchmark::State&)`.